File: packages/node/src/client/daemon-client.ts
import type { ServerNotification } from '@modelcontextprotocol/sdk/types.js';
import { portLang, t } from '../i18n/lang.js';
import { ClientErrorCode } from '../types/errors.js';

export type DaemonToolContent = { type: 'text'; text: string };

export interface DaemonCallResult {
  content: DaemonToolContent[];
  isError?: boolean;
}

export interface CallDaemonOptions {
  authToken?: string;
  progressToken?: string | number;
  onProgress?: (notification: ServerNotification) => Promise<void>;
  signal?: AbortSignal;
}

export interface DaemonHealthOk {
  ok: true;
  identity?: string;
  bridgePort: number;
  sessionCount: number;
}

function authHeaders(token: string | undefined): Record<string, string> {
  if (!token) return {};
  return { Authorization: `Bearer ${token}` };
}

function daemonUnreachableCallResult(cause?: string): DaemonCallResult {
  const L = portLang();
  return {
    content: [
      {
        type: 'text',
        text: JSON.stringify({
          ok: false,
          error: {
            code: ClientErrorCode.DAEMON_UNREACHABLE,
            message: t(L, 'daemon_unreachable.message'),
            details: {
              startCommand: 'lionscraper daemon',
              hint: t(L, 'daemon_unreachable.hint'),
              ...(cause ? { cause } : {}),
            },
          },
        }),
      },
    ],
    isError: true,
  };
}

function invalidDaemonResponseResult(message: string, details?: Record<string, unknown>): DaemonCallResult {
  return {
    content: [
      {
        type: 'text',
        text: JSON.stringify({
          ok: false,
          error: {
            code: ClientErrorCode.DAEMON_INVALID_RESPONSE,
            message,
            ...(details !== undefined && Object.keys(details).length > 0 ? { details } : {}),
          },
        }),
      },
    ],
    isError: true,
  };
}

async function processNdjsonLineArray(
  lineChunks: string[],
  onProgress: (notification: ServerNotification) => Promise<void>,
): Promise<DaemonCallResult> {
  let final: DaemonCallResult | null = null;

  for (const line of lineChunks) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    let obj: {
      type: string;
      notification?: ServerNotification;
      content?: DaemonToolContent[];
      isError?: boolean;
      error?: { message?: string };
    };
    try {
      obj = JSON.parse(trimmed) as typeof obj;
    } catch {
      return invalidDaemonResponseResult('Invalid JSON line in daemon NDJSON stream', {
        linePreview: trimmed.slice(0, 120),
      });
    }
    if (obj.type === 'progress' && obj.notification) {
      await onProgress(obj.notification);
    } else if (obj.type === 'result' && Array.isArray(obj.content)) {
      final = { content: obj.content, isError: obj.isError };
    } else if (obj.type === 'error') {
      return invalidDaemonResponseResult(obj.error?.message ?? 'Daemon tool error line in NDJSON stream');
    }
  }

  if (!final) {
    return invalidDaemonResponseResult('No result line in daemon NDJSON stream (incomplete response)');
  }
  return final;
}

async function readNdjsonStream(
  res: Response,
  onProgress: (notification: ServerNotification) => Promise<void>,
): Promise<DaemonCallResult> {
  if (!res.body) {
    return invalidDaemonResponseResult('Empty response body from daemon (streaming)');
  }
  const reader = res.body.pipeThrough(new TextDecoderStream()).getReader();
  let buffer = '';
  const completeLines: string[] = [];

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += value;
    const lines = buffer.split('\n');
    buffer = lines.pop() ?? '';
    for (const line of lines) {
      completeLines.push(line);
    }
  }
  if (buffer.trim()) {
    completeLines.push(buffer);
  }

  return processNdjsonLineArray(completeLines, onProgress);
}

/** When Content-Type is not ndjson but body may still be NDJSON lines (misconfigured proxy, etc.). */
async function tryReadNdjsonFromTextBody(
  text: string,
  onProgress: (notification: ServerNotification) => Promise<void>,
): Promise<DaemonCallResult | null> {
  const rawLines = text.split('\n').filter((l) => l.trim());
  if (rawLines.length === 0) return null;
  let looksNdjson = false;
  for (const l of rawLines) {
    try {
      const o = JSON.parse(l) as { type?: string };
      if (o.type === 'progress' || o.type === 'result' || o.type === 'error') {
        looksNdjson = true;
        break;
      }
    } catch {
      /* ignore */
    }
  }
  if (!looksNdjson) return null;
  return processNdjsonLineArray(rawLines, onProgress);
}

/**
 * POST /v1/tools/call on the LionScraper daemon HTTP control plane.
 * Network failures are returned as {@link ClientErrorCode.DAEMON_UNREACHABLE} (not thrown).
 * Malformed streaming or JSON bodies use {@link ClientErrorCode.DAEMON_INVALID_RESPONSE}.
 */
export async function callDaemonTool(
  baseUrl: string,
  name: string,
  args: Record<string, unknown>,
  options: CallDaemonOptions = {},
): Promise<DaemonCallResult> {
  try {
    return await callDaemonToolUnsafe(baseUrl, name, args, options);
  } catch (err) {
    const cause = err instanceof Error ? err.message : String(err);
    return daemonUnreachableCallResult(cause);
  }
}

async function callDaemonToolUnsafe(
  baseUrl: string,
  name: string,
  args: Record<string, unknown>,
  options: CallDaemonOptions,
): Promise<DaemonCallResult> {
  const url = `${baseUrl.replace(/\/$/, '')}/v1/tools/call`;
  const useStream =
    options.progressToken !== undefined && options.progressToken !== null && options.onProgress !== undefined;

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...authHeaders(options.authToken),
  };
  if (useStream) {
    headers.Accept = 'application/x-ndjson';
  }

  const body = JSON.stringify({
    name,
    arguments: args,
    ...(useStream ? { progressToken: options.progressToken } : {}),
  });

  const res = await fetch(url, {
    method: 'POST',
    headers,
    body,
    signal: options.signal,
  });

  if (useStream && res.ok) {
    const ct = (res.headers.get('content-type') ?? '').toLowerCase();
    if (ct.includes('ndjson')) {
      return readNdjsonStream(res, options.onProgress!);
    }
    const text = await res.text();
    const fromNdjson = await tryReadNdjsonFromTextBody(text, options.onProgress!);
    if (fromNdjson) return fromNdjson;
    try {
      return JSON.parse(text) as DaemonCallResult;
    } catch {
      return invalidDaemonResponseResult(
        'Expected application/x-ndjson from daemon for streaming tool call, but body was not valid NDJSON or JSON',
        { contentType: res.headers.get('content-type') ?? '', snippet: text.slice(0, 200) },
      );
    }
  }

  const text = await res.text();
  if (!res.ok) {
    if (res.status >= 500) {
      return daemonUnreachableCallResult(text || `HTTP ${res.status}`);
    }
    try {
      const j = JSON.parse(text) as { error?: { message?: string }; ok?: boolean };
      const msg = j.error?.message ?? (text || `HTTP ${res.status}`);
      return {
        content: [{ type: 'text', text: JSON.stringify({ ok: false, error: { message: msg, status: res.status } }) }],
        isError: true,
      };
    } catch {
      return {
        content: [{ type: 'text', text: text || `HTTP ${res.status}` }],
        isError: true,
      };
    }
  }

  try {
    return JSON.parse(text) as DaemonCallResult;
  } catch {
    return invalidDaemonResponseResult('Invalid JSON from daemon (non-streaming response)', {
      snippet: text.slice(0, 200),
    });
  }
}

/**
 * GET /v1/health — requires a LionScraper daemon with WebSocket bridge ready (`bridgePort` > 0).
 */
export async function daemonHealth(
  baseUrl: string,
  authToken?: string,
  signal?: AbortSignal,
): Promise<DaemonHealthOk> {
  const url = `${baseUrl.replace(/\/$/, '')}/v1/health`;
  let res: Response;
  try {
    res = await fetch(url, {
      method: 'GET',
      headers: { ...authHeaders(authToken) },
      signal,
    });
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    throw new Error(msg);
  }

  const text = await res.text();
  if (!res.ok) {
    throw new Error(text || `HTTP ${res.status}`);
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw new Error('Invalid JSON from daemon health endpoint');
  }

  if (typeof parsed !== 'object' || parsed === null) {
    throw new Error('Invalid daemon health response');
  }

  const o = parsed as Record<string, unknown>;
  if (o.ok !== true) {
    throw new Error('Daemon health reports not ready');
  }

  if (o.identity !== undefined && o.identity !== 'lionscraper') {
    throw new Error('Not a LionScraper daemon (identity mismatch)');
  }

  const bridgePort = o.bridgePort;
  if (typeof bridgePort !== 'number' || !Number.isFinite(bridgePort) || bridgePort <= 0) {
    throw new Error('Daemon WebSocket bridge is not ready (invalid bridgePort)');
  }

  const rawSc = o.sessionCount;
  const sessionCount =
    typeof rawSc === 'number' && Number.isFinite(rawSc) ? Math.max(0, Math.floor(rawSc)) : 0;

  return {
    ok: true,
    bridgePort,
    sessionCount,
    ...(typeof o.identity === 'string' ? { identity: o.identity } : {}),
  };
}

File: packages/python/src/lionscraper/client/daemon_client.py
from __future__ import annotations

import json
from typing import Any, Awaitable, Callable, TypedDict

import aiohttp

from lionscraper.i18n.lang import port_lang, t
from lionscraper.types.errors import ClientErrorCode


class DaemonCallResult(TypedDict, total=False):
    content: list[dict[str, Any]]
    isError: bool


def _auth_headers(token: str | None) -> dict[str, str]:
    if not token:
        return {}
    return {"Authorization": f"Bearer {token}"}


def _daemon_unreachable_result(cause: str | None = None) -> DaemonCallResult:
    L = port_lang()
    details: dict[str, Any] = {
        "startCommand": "lionscraper daemon",
        "hint": t(L, "daemon_unreachable.hint"),
    }
    if cause:
        details["cause"] = cause
    return {
        "content": [
            {
                "type": "text",
                "text": json.dumps(
                    {
                        "ok": False,
                        "error": {
                            "code": ClientErrorCode.DAEMON_UNREACHABLE.value,
                            "message": t(L, "daemon_unreachable.message"),
                            "details": details,
                        },
                    },
                    ensure_ascii=False,
                ),
            }
        ],
        "isError": True,
    }


def _invalid_response_result(message: str, details: dict[str, Any] | None = None) -> DaemonCallResult:
    err: dict[str, Any] = {
        "code": ClientErrorCode.DAEMON_INVALID_RESPONSE.value,
        "message": message,
    }
    if details:
        err["details"] = details
    return {
        "content": [{"type": "text", "text": json.dumps({"ok": False, "error": err}, ensure_ascii=False)}],
        "isError": True,
    }


async def _process_ndjson_lines(
    lines: list[str],
    on_progress: Callable[[dict[str, Any]], Awaitable[None]],
) -> DaemonCallResult:
    final: DaemonCallResult | None = None
    for line in lines:
        trimmed = line.strip()
        if not trimmed:
            continue
        try:
            obj: dict[str, Any] = json.loads(trimmed)
        except json.JSONDecodeError:
            return _invalid_response_result(
                "Invalid JSON line in daemon NDJSON stream",
                {"linePreview": trimmed[:120]},
            )
        if obj.get("type") == "progress" and obj.get("notification"):
            await on_progress(obj["notification"])
        elif obj.get("type") == "result" and isinstance(obj.get("content"), list):
            final = {"content": obj["content"], "isError": bool(obj.get("isError"))}
        elif obj.get("type") == "error":
            return _invalid_response_result(
                str((obj.get("error") or {}).get("message") or "Daemon tool error line in NDJSON stream")
            )
    if not final:
        return _invalid_response_result("No result line in daemon NDJSON stream (incomplete response)")
    return final


def _handle_non_ok_response(status: int, text: str) -> DaemonCallResult:
    if status >= 500:
        return _daemon_unreachable_result(text or f"HTTP {status}")
    try:
        j = json.loads(text)
        msg = j.get("error", {}).get("message") or text or f"HTTP {status}"
        return {
            "content": [
                {
                    "type": "text",
                    "text": json.dumps({"ok": False, "error": {"message": msg, "status": status}}, ensure_ascii=False),
                }
            ],
            "isError": True,
        }
    except json.JSONDecodeError:
        return {"content": [{"type": "text", "text": text or f"HTTP {status}"}], "isError": True}


async def _try_read_ndjson_from_text_body(
    text: str,
    on_progress: Callable[[dict[str, Any]], Awaitable[None]],
) -> DaemonCallResult | None:
    raw_lines = [l for l in text.split("\n") if l.strip()]
    if not raw_lines:
        return None
    looks = False
    for l in raw_lines:
        try:
            o = json.loads(l)
            if o.get("type") in ("progress", "result", "error"):
                looks = True
                break
        except json.JSONDecodeError:
            pass
    if not looks:
        return None
    return await _process_ndjson_lines(raw_lines, on_progress)


async def call_daemon_tool(
    base_url: str,
    name: str,
    args: dict[str, Any],
    *,
    auth_token: str | None = None,
    progress_token: str | int | None = None,
    on_progress: Callable[[dict[str, Any]], Awaitable[None]] | None = None,
) -> DaemonCallResult:
    try:
        return await _call_daemon_tool_unsafe(
            base_url, name, args, auth_token, progress_token, on_progress
        )
    except Exception as e:
        return _daemon_unreachable_result(str(e) if e else None)


async def _call_daemon_tool_unsafe(
    base_url: str,
    name: str,
    args: dict[str, Any],
    auth_token: str | None,
    progress_token: str | int | None,
    on_progress: Callable[[dict[str, Any]], Awaitable[None]] | None,
) -> DaemonCallResult:
    url = f"{base_url.rstrip('/')}/v1/tools/call"
    use_stream = progress_token is not None and on_progress is not None
    headers: dict[str, str] = {"Content-Type": "application/json", **_auth_headers(auth_token)}
    if use_stream:
        headers["Accept"] = "application/x-ndjson"
    body: dict[str, Any] = {"name": name, "arguments": args}
    if use_stream:
        body["progressToken"] = progress_token

    timeout = aiohttp.ClientTimeout(total=600.0)
    async with aiohttp.ClientSession(timeout=timeout) as session:
        if use_stream:
            async with session.post(url, headers=headers, json=body) as res:
                text = await res.text(encoding="utf-8", errors="replace")
                if res.status == 200:
                    ct = (res.headers.get("Content-Type") or "").lower()
                    lines = text.split("\n")
                    if "ndjson" in ct:
                        return await _process_ndjson_lines(lines, on_progress)  # type: ignore[arg-type]
                    from_ndjson = await _try_read_ndjson_from_text_body(text, on_progress)  # type: ignore[arg-type]
                    if from_ndjson:
                        return from_ndjson
                    try:
                        return json.loads(text)
                    except json.JSONDecodeError:
                        return _invalid_response_result(
                            "Expected application/x-ndjson from daemon for streaming tool call, but body was not valid NDJSON or JSON",
                            {"contentType": res.headers.get("Content-Type") or "", "snippet": text[:200]},
                        )
                return _handle_non_ok_response(res.status, text)

        async with session.post(url, headers=headers, json=body) as res:
            text = await res.text(encoding="utf-8", errors="replace")
            if res.status != 200:
                return _handle_non_ok_response(res.status, text)
            try:
                return json.loads(text)
            except json.JSONDecodeError:
                return _invalid_response_result(
                    "Invalid JSON from daemon (non-streaming response)",
                    {"snippet": text[:200]},
                )


async def daemon_health(base_url: str, auth_token: str | None = None) -> dict[str, Any]:
    url = f"{base_url.rstrip('/')}/v1/health"
    timeout = aiohttp.ClientTimeout(total=10.0)
    try:
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(url, headers=_auth_headers(auth_token)) as res:
                text = await res.text(encoding="utf-8", errors="replace")
                if res.status < 200 or res.status >= 300:
                    raise RuntimeError(text or f"HTTP {res.status}")
                try:
                    parsed = json.loads(text)
                except json.JSONDecodeError as e:
                    raise RuntimeError("Invalid JSON from daemon health endpoint") from e
                if not isinstance(parsed, dict):
                    raise RuntimeError("Invalid daemon health response")
                if parsed.get("ok") is not True:
                    raise RuntimeError("Daemon health reports not ready")
                if parsed.get("identity") is not None and parsed.get("identity") != "lionscraper":
                    raise RuntimeError("Not a LionScraper daemon (identity mismatch)")
                bp = parsed.get("bridgePort")
                if not isinstance(bp, (int, float)) or bp <= 0:
                    raise RuntimeError("Daemon WebSocket bridge is not ready (invalid bridgePort)")
                raw_sc = parsed.get("sessionCount", 0)
                session_count = max(0, int(raw_sc)) if isinstance(raw_sc, (int, float)) else 0
                out: dict[str, Any] = {"ok": True, "bridgePort": int(bp), "sessionCount": session_count}
                if isinstance(parsed.get("identity"), str):
                    out["identity"] = parsed["identity"]
                impl = parsed.get("implementation")
                if isinstance(impl, str) and impl in ("node", "python"):
                    out["implementation"] = impl
                return out
    except aiohttp.ClientError as e:
        raise RuntimeError(str(e)) from e
